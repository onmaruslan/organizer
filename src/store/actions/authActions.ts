import { Dispatch } from "redux"
import firebase from 'firebase/app'
import 'firebase/auth';        // for authentication

import {AuthAction, AuthActionTypes} from '../types/authTypes'

export function changeUser(user: object | undefined | boolean): AuthAction {
  return { type: AuthActionTypes.CHANGE_USER, payload: user }
}

export const setError = (error: null | string): AuthAction => {
  return {type: AuthActionTypes.ERROR, payload: error}
}
export const setSuccess = (success: null | string): AuthAction => {
  return {type: AuthActionTypes.SUCCESS, payload: success}
}

export const signIn = (email: string, password: string, remember: boolean) => {
  return (dispatch: Dispatch<AuthAction>) => {
    const session = (remember) ? 'LOCAL' : 'SESSION'
    firebase.auth().setPersistence(firebase.auth.Auth.Persistence[session])
      .then(() => {
        return firebase.auth().signInWithEmailAndPassword(email, password)
          .then((userCredential: any) => {
            console.log(userCredential.user)
            dispatch(changeUser(userCredential.user));

            (remember)
              ? localStorage.setItem('user', JSON.stringify(userCredential.user))
              : sessionStorage.setItem('user', JSON.stringify(userCredential.user))

          })
      })
      .catch((error) => {
        dispatch(setError(error.code))
        switch (error.code) {
          case 'auth/user-not-found':
            dispatch(setError('Email или пароль введены не верно'))
            break;
          case 'auth/wrong-password':
            dispatch(setError('Email или пароль введены не верно'))
            break;
          default:
            dispatch(setError(error.message))
        }
      });

  }
}

export const signUp = (email: string, password: string) => {
  return (dispatch: Dispatch<AuthAction>) => {
    firebase.auth().createUserWithEmailAndPassword(email, password)
      .then((userCredential: any) => {
        dispatch(changeUser(userCredential.user))
        localStorage.setItem('user', JSON.stringify(userCredential.user))
      })
      .catch((error) => {
        switch (error.code) {
          case 'auth/email-already-in-use':
            dispatch(setError('Такой email уже используется'))
            break;
          default:
            dispatch(setError(error.message))
        }
      })
  }
}

export const sendPasswordReset = (email: string) => {
  return (dispatch: Dispatch<AuthAction>) => {
    firebase.auth().sendPasswordResetEmail(email).then(function() {
      dispatch(setSuccess(`Письмо отправлено на почту ${email}`))

    }).catch(function(error) {
      switch (error.code) {
        case 'auth/user-not-found':
          dispatch(setError('Такой пользователь не зарегистрирован'))
          break;
        default:
          dispatch(setError(error.message))
      }
    });
  }

}


export const signOut = () => {
  return (dispatch: Dispatch<AuthAction>) => {

    firebase
      .auth()
      .signOut()
      .then(() => {
        dispatch(changeUser(false))
        localStorage.clear()
        sessionStorage.clear()
      })
  }
}