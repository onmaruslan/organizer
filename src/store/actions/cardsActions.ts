import { Dispatch } from "redux"
import firebase from 'firebase/app'
import {HomeAction, HomeActionTypes} from '../types/homeTypes'


export function addCard(): HomeAction {
  return { type: HomeActionTypes.ADD_CARD , payload: Date.now()}
}
export function addTodo(idCard: number): HomeAction {
  return { type: HomeActionTypes.ADD_TODO , payload: {idCard: idCard, idTodo: Date.now()}}
}