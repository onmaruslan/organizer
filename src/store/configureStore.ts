import { applyMiddleware, combineReducers, createStore } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension'
import thunk from 'redux-thunk'

import {authReducer} from './reducers/authReducer'
import {cardsReducer} from './reducers/cardsReducer'

const rootReducer = combineReducers({
    auth: authReducer,
    cards: cardsReducer
})
export type RootState = ReturnType<typeof rootReducer>

export const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)))