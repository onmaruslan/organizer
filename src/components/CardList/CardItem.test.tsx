import React from 'react'
import { Provider } from 'react-redux'
import renderer, { ReactTestRenderer } from 'react-test-renderer'
import configureStore, {MockStore} from 'redux-mock-store'

import CardItem from "./CardItem"
import {removeTodo, setTitleTodo} from '../../store/actions/cardsActions'

const mockStore = configureStore([])

describe('CardItem Component', () => {
  let store : MockStore
  let component : ReactTestRenderer

  beforeEach(() => {
    store = mockStore()
    store.dispatch = jest.fn()

    component = renderer.create(
      <Provider store={store}>
        <CardItem todoDB={{title: '123', id: 1}} cardID={0}/>
      </Provider>
    )
  })

  it('should render with given state from props', () => {
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should dispatch an action on input change', () => {
    renderer.act(() => {
      component.root.findByType('input')
        .props.onChange({ target: { value: 'New Title' } })

      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(
        setTitleTodo(0, 1, 'New Title')
      )
    })
  })

  it('should dispatch an action on removeButton click', () => {
    renderer.act(() => {
      component.root.findByProps({className: "card__item_delete"}).props.onClick()

      expect(store.dispatch).toHaveBeenCalledTimes(1)
      expect(store.dispatch).toHaveBeenCalledWith(
        removeTodo(0, 1)
      )
    })
  })
})