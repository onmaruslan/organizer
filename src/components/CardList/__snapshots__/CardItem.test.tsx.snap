// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem Component should render with given state from props 1`] = `
Array [
  <input
    onChange={[Function]}
    placeholder="Новая задача"
    type="text"
    value="123"
  />,
  <div
    className="card__item_shadow"
  />,
  <div
    className="card__item_position"
    data-movable-handle={true}
    title="Изменить расположение задачи"
  >
    <svg
      viewBox="0 0 512 512"
    >
      <path
        d="M31.9 125.9h73.8V52.2H31.9v73.7zm84.1 20.9H21.4c-5.8 0-10.4-4.7-10.4-10.4V41.7c0-5.8 4.7-10.4 10.4-10.4H116c5.8 0 10.4 4.7 10.4 10.4v94.6c.1 5.8-4.6 10.5-10.4 10.5zm374.6-47.3H166.1c-5.8 0-10.4-4.7-10.4-10.4 0-5.8 4.7-10.4 10.4-10.4h324.5c5.8 0 10.4 4.7 10.4 10.4s-4.7 10.4-10.4 10.4zM31.9 292.9h73.8v-73.8H31.9v73.8zm84.1 20.8H21.4c-5.8 0-10.4-4.7-10.4-10.4v-94.6c0-5.8 4.7-10.4 10.4-10.4H116c5.8 0 10.4 4.7 10.4 10.4v94.6c.1 5.8-4.6 10.4-10.4 10.4zm374.6-47.3H166.1c-5.8 0-10.4-4.7-10.4-10.4 0-5.8 4.7-10.4 10.4-10.4h324.5c5.8 0 10.4 4.7 10.4 10.4 0 5.8-4.7 10.4-10.4 10.4zM31.9 459.8h73.8V386H31.9v73.8zm84.1 20.9H21.4c-5.8 0-10.4-4.7-10.4-10.4v-94.6c0-5.8 4.7-10.4 10.4-10.4H116c5.8 0 10.4 4.7 10.4 10.4v94.6c.1 5.7-4.6 10.4-10.4 10.4zm374.6-47.3H166.1c-5.8 0-10.4-4.7-10.4-10.4s4.7-10.4 10.4-10.4h324.5c5.8 0 10.4 4.7 10.4 10.4s-4.7 10.4-10.4 10.4z"
      />
    </svg>
  </div>,
  <div
    className="card__item_delete"
    onClick={[Function]}
    title="Удалить задачу"
  >
    <svg
      viewBox="0 0 240 240"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M120 240c66.168 0 120-53.831 120-120S186.168 0 120 0 0 53.832 0 120s53.832 120 120 120zm0-210c49.626 0 90 40.374 90 90s-40.374 90-90 90-90-40.374-90-90 40.374-90 90-90zM69.144 149.644L98.787 120 69.144 90.356l21.213-21.213L120 98.787l29.644-29.644 21.213 21.213L141.213 120l29.643 29.644-21.213 21.213L120 141.213l-29.644 29.643-21.212-21.212z"
      />
    </svg>
  </div>,
]
`;
